bundles/org.openhab.binding.somfymylink/src/main/java/org/openhab/binding/somfymylink/internal/handler/SomfyMyLinkBridgeHandler.java

```diff
@@ -0,0 +1,452 @@
+                        while (((readCount = in.read(readBuff)) != -1)) {
+                            message.append(new String(readBuff, 0, readCount));
+
+                            // try and parse the message
+                            try {
+                                T response = parseResponse(message.toString(), responseType);
+                                logger.debug("Got full message: {}", message);
+                                future.complete(response);
+                                return;
+                            } catch (SomfyMyLinkException e) {
+                                // trouble parsing the message, probably incomplete
+                                // if its been over 1 min waiting for a correct and full response then abort
+                                elapsedTime = (new Date()).getTime() - startTime;
+                                if (elapsedTime >= 60000) {
+                                    future.completeExceptionally(
+                                            new SomfyMyLinkException("Timeout waiting for reply from mylink"));
+                                    return;
+                                }
+                                Thread.sleep(250); // pause for 250ms
+                            }
+                        }
+                    }
+
+                    // only if we didn't already get a response give time for mylink to process
+                    Thread.sleep(CONNECTION_DELAY);
+
+                    future.complete(null);
+                } catch (SocketTimeoutException e) {
+                    logger.warn("Timeout sending command to mylink: {} Message: {}", json, e.getMessage());
+                    future.completeExceptionally(new SomfyMyLinkException("Timeout sending command to mylink", e));
+                } catch (IOException e) {
+                    logger.warn("Problem sending command to mylink: {} Message: {}", json, e.getMessage());
+                    future.completeExceptionally(new SomfyMyLinkException("Problem sending command to mylink", e));
+                } catch (InterruptedException e) {
+                    logger.warn("Interrupted while waiting after sending command to mylink: {} Message: {}", json,
+                            e.getMessage());
+                    future.complete(null);
+                } catch (Exception e) {
+                    logger.warn("Unexpected exception while sending command to mylink: {} Message: {}", json,
+                            e.getMessage());
+                    future.completeExceptionally(e);
+                }
+            });
+        } else {
+            future.complete(null);
+        }
+        return future;
+    }
+
+    private <T extends SomfyMyLinkResponseBase> T parseResponse(String message, Class<T> responseType) {
+        JsonParser parser = new JsonParser();
+        JsonObject jsonObj = parser.parse(message).getAsJsonObject();
```

Comment:
K, from what you have showed it looks like the somfy devices are not closing the stream after they finish writing the json to it. The reason your buffered stream solution works is that it will only fill the buffer with data that is immediately available.

The reason that my previous gson suggestion didn't work is that passing a reader directly to a JsonParser makes the parser check for additional tokens after it finishes parsing a json object. But apparently it doesn't do the same check when you pass in a JsonReader.
So the following code change should work. If not, I give up.
```suggestion
    private <T extends SomfyMyLinkResponseBase> T parseResponse(Reader reader, Class<T> responseType) {
        JsonParser parser = new JsonParser();
        JsonObject jsonObj = parser.parse(gson.newJsonReader(reader)).getAsJsonObject();
```

Reply:
success! this worked. switched to it & also to a Writer for writing to socket and removed all char[] buffer code.

Reply:
@cpmeister Should a new `JsonParser` be created every time a message is parsed? In the past, I've been given guidance to create a singleton `Gson` instance rather than create one every time I needed to use it. Wondering if the same is true for `JsonParser`. WDYT?

@LoungeFlyZ Looking good. If you want some additional testing of the latest version, please point me to a jar file.

Reply:
@mhilbush here is a new jar if you would like to give it a whirl! https://www.dropbox.com/s/udc4s3tqv5zfviy/org.openhab.binding.somfymylink-2.5.5-SNAPSHOT.jar?dl=1

Reply:
@mhilbush a `Gson` object is a heavy object with dozens of settings fields which would be costly to reconfigure from scratch every time. On the other hand a `JsonParser` object has no fields to speak of, it is stateless and could easily be turned into a static utility class (why the didn't is beyond me). So the cost of creating a new JsonParser each time is actually less than the cost of creating lambdas.

In general though, it is good to reuse objects when you can but for something as light as JsonParser it probably isn't worth the overhead of adding another field.

Reply:
Thanks. Very good to know.